v2/gocli/go_binary.go

```diff
@@ -0,0 +1,111 @@
+// limitations under the License.
+
+package gocli
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/go-licenses/v2/third_party/uw-labs/lichen/model"
+	lichenmodule "github.com/google/go-licenses/v2/third_party/uw-labs/lichen/module"
+)
+
+// Module metadata extracted from binary and local go module workspace.
+type BinaryMetadata struct {
+	// The main module used to build the binary.
+	// e.g. github.com//google/go-licenses/v2/tests/modules/cli02
+	MainModule string
+	// Import path of the main package to build the binary.
+	// e.g. github.com//google/go-licenses/v2/tests/modules/cli02/cmd
+	Path string
+	// Detailed metadata of all the module dependencies.
+	// Does not include the main module.
+	Modules []Module
+}
+
+// List dependencies from module metadata in a go binary.
+//
+// Prerequisites:
+// * The go binary must be built with go modules without any further modifications.
+// * The command must run with working directory same as to build the analyzed
+// go binary, because we need the exact go modules info used to build it.
+//
+// Here, I am using [1] as a short term solution. It runs [4] go version -m and parses
+// output. This is preferred over [2], because [2] is an alternative implemention
+// for go version -m, and I expect better long term compatibility for go version -m.
+//
+// The parsing command output hack is still unfavorable in the long term. As
+// dicussed in [3], golang community will move go version parsing into an individual
+// module in golang.org/x. We can use that module instead after it is built.
+//
+// References of similar implementations or dicussions:
+// 1. https://github.com/uw-labs/lichen/blob/be9752894a5958f6ba7be9e05dc370b7a73b58db/internal/module/extract.go#L16
+// 2. https://github.com/mitchellh/golicense/blob/8c09a94a11ac73299a72a68a7b41e3a737119f91/module/module.go#L27
+// 3. https://github.com/golang/go/issues/39301
+// 4. https://golang.org/pkg/cmd/go/internal/version/
+func ExtractBinaryMetadata(path string) (*BinaryMetadata, error) {
+	buildInfo, err := listModulesInBinary(path)
+	if err != nil {
+		return nil, err
+	}
+	mods, err := joinModulesMetadata(buildInfo.ModuleRefs)
+	if err != nil {
+		return nil, err
+	}
+	return &BinaryMetadata{
+		MainModule: buildInfo.ModulePath,
+		Path:       buildInfo.PackagePath,
+		Modules:    mods,
+	}, nil
+}
+
+func listModulesInBinary(path string) (buildinfo *model.BuildInfo, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("listModulesInGoBinary(path=%q): %w", path, err)
+		}
+	}()
+	depsBuildInfo, err := lichenmodule.Extract(context.Background(), path)
+	if err != nil {
+		return nil, err
+	}
+	if len(depsBuildInfo) != 1 {
+		return nil, fmt.Errorf("len(depsBuildInfo) should be 1, but found %v", len(depsBuildInfo))
+	}
+	return &depsBuildInfo[0], nil
+}
+
+func joinModulesMetadata(refs []model.ModuleReference) (modules []Module, err error) {
+	localModules, err := ListModules()
```

Comment:
Yes, I am taking that as a prerequisite. It works for my use case, because I built the binary, so I always have access to the module dir. I am curious when you wouldn't.

I see your point for making it also accept package path, so it could be backwards compatible. I think that's a good idea. I will put it on the roadmap as a P1. I prefer getting the end to end flow working before introducing options. What do you think?

I didn't quite understand one point:
> since we could go from binary path -> module path -> packages

How do we get packages from module path if we are not already in the module dir? Do you mean scenarios like the module itself lives in GOPATH?

Reply:
Added in roadmap: https://github.com/google/go-licenses/issues/70#issuecomment-864388550

Reply:
2 use cases come to mind -

1. CI jobs that pull in multiple repos - the modules would be in subdirectories and might not start from a module directory.
2. If you want to query for just the licenses of the binary (without the module).

I think it's clearer to have users specify `go-licenses check path/to/binary ./...`. I'm worried that baking in the assumption that users will always execute the tool from a module directory may not work for all use cases.

Reply:
Ah - I just meant that for a given arg we can try to parse it as a binary first, get the modules, then use the same packages.Load that we're already using to look up the deps.

Reply:
I think there are some complexities in getting the same information outside of the go module, which makes me prefer not supporting this feature initially. Let me confirm my understanding, when outside of a go module, we can only resolve packages in GOPATH mode, so is "packages.Load". Is that right? Therefore, whenever possible I think people should avoid doing that.

Complexities:
1. versioning, outside of the go module, we cannot go get a module with version. (major concern)
2. [replace directives](https://golang.org/ref/mod#go-mod-file-replace), a replaced module may be a local path too, in which case, you can only know what code exactly was when you are checking licenses in the go module. (probably not a strong argument, because this is rather corner case.)

> 2 use cases come to mind -
> 
> 1. CI jobs that pull in multiple repos - the modules would be in subdirectories and might not start from a module directory.

Because of complexities mentioned above, I think for the same reasons, to run commands like `go build`, `go test`, you would also need to cd into the module directory. I feel like it's fair to assume we need the same prerequisite when checking licenses.

But I think you reminded me that, we do not need to require current working dir is the root of the module. We should be able to run it in sub folders just like other go CLI commands.

> 2. If you want to query for just the licenses of the binary (without the module).
> 
> I think it's clearer to have users specify `go-licenses check path/to/binary ./...`. I'm worried that baking in the assumption that users will always execute the tool from a module directory may not work for all use cases.

I think this would be a useful scenario, just that I cannot understand how we can download source code of all dependencies without the module directory?
I can think of probably two ways:
1. resolve from GOPATH -- this also assumes sth on user environment and there can be version skew. So I'd refrain from doing this
2. construct a temporary go module with a go.mod including all the modules we found in the binary, and then use the dummy go module env to download all source code and check dependencies. This probably works, but seems hacky too.

Reply:
Let me confirm, do you think this is a blocker for this PR?

Reply:
Did some testing of the existing behavior - looks like this is influenced by the `GO111MODULE` environment variable. 

You're right that even with the GOPATH mode, we still need to run within the GOPATH for things to work, and that this is consistent with other go tools, and even the existing go-licenses tool (with `GO111MODULE=on` go-licenses needs to run within a module directory to function properly). So LGTM for requiring this.

I still want to push back on automatically including everything in the current module directory by default. In the same way you need to distinguish check 1 package (`go-licenses check .`) vs check all packages (`go-licenses ./...`), I'd like to do something similar for binaries as well. Something else I realized in messing with things is that `go version` already recursively traverses directories for binaries, so I'm wondering how we can make a better distinction between the package `.` vs "look for binaries in this directory" `.`. 2 possible ideas -

1. Put binary commands into their own command group similar to `go list` vs `go version`
  
  ```sh
  $ go-licenses binary save .
  ```

2. Put binary search path into a flag

  ```sh
  $ go-licenses binary save --binary_path=. ./...
  ```

Let me know what you think!

Reply:
> So LGTM for requiring this.

Great, we agreed on one more thing.

> I still want to push back on automatically including everything in the current module directory by default.

I guess my function naming can be a little misleading? `join` --> `inner join`.
I am already doing what you mentioned -- reading all modules in current workdir, but I do not add all of them by default. Only modules used in the go binary are returned.

> Something else I realized in messing with things is that go version already recursively traverses directories for binaries, so I'm wondering how we can make a better distinction between the package . vs "look for binaries in this directory" .

Interesting, so far my use-case is limited to getting licenses for one binary. Let me think about what's a better way to handle the ambiguity.

Anyway, I think this PR only implements listing modules in a binary. We can keep discussing what's the best CLI commands + semantics in upcoming PRs.